src/ray/util/util.h

```diff
@@ -237,3 +238,64 @@ inline void SetThreadName(const std::string &thread_name) {
   pthread_setname_np(pthread_self(), thread_name.substr(0, 15).c_str());
 #endif
 }
+
+inline std::string GetThreadName() {
+#if defined(__linux__)
+  char name[128];
+  auto rc = pthread_getname_np(pthread_self(), name, sizeof(name));
+  if (rc != 0) {
+    return "ERROR";
+  } else {
+    return name;
+  }
+#else
+  return "UNKNOWN";
+#endif
+}
+
+namespace ray {
+template <typename T>
+class ThreadIdempotent {
+ public:
+  template <typename... Ts>
+  ThreadIdempotent(Ts &&... ts) : t_(std::forward<Ts>(ts)...) {}
+
+  T &operator*() {
+    ThreadCheck();
+    return t_;
+  }
+
+  T *operator->() {
+    ThreadCheck();
+    return &t_;
+  }
+
+  const T &operator*() const {
+    ThreadCheck();
+    return t_;
+  }
+
+  const T *operator->() const {
+    ThreadCheck();
+    return &t_;
+  }
+
+ private:
+  void ThreadCheck() const {
+    if (id_ == std::thread::id()) {
+      thread_name_ = GetThreadName();
+      RAY_LOG(DEBUG) << "Run in thread: " << thread_name_;
```

Comment:
```suggestion
      RAY_LOG(DEBUG) << "Created in thread: " << thread_name_;
```

Reply:
It's actually not creating, it's the first time it's used. How about 'Accessed in thread: ..'

Reply:
Oh, that's a bit weird. Why not first time created?